File: framework/python/doppel/policy.py
from dataclasses import dataclass
import re

from .models import Command, Observation


@dataclass(frozen=True)
class Verdict:
    decision: str
    reason: str
    human_takeover: str | None = None


class ActionPolicy:
    SENSITIVE_WORDS = ("发送", "删除", "提交订单", "确认下单", "send", "delete", "submit order")
    VERIFICATION_HEADING = re.compile(r"^(?:安全验证|人机验证|图形验证|滑块验证|验证码安全校验|security verification|human verification|verify (?:that )?you are (?:a )?human|i['’]?m not a robot|captcha|recaptcha|hcaptcha)[.!。！]?$", re.IGNORECASE)
    VERIFICATION_INSTRUCTION = re.compile(r"(?:请|需要).{0,12}(?:完成|进行).{0,12}(?:人机|安全)验证|请验证[您你]是真人|(?:拖动|滑动).{0,12}滑块.{0,20}(?:验证|拼图)|(?:drag|slide).{0,24}slider.{0,40}(?:verif|puzzle)|(?:select|click) all (?:images|squares) (?:with|containing)|^verify (?:that )?you are (?:a )?human[.!]?$|^i['’]?m not a robot[.!]?$", re.IGNORECASE)

    def requires_verification(self, observation: Observation | None) -> bool:
        if observation is None:
            return False
        return any(self.VERIFICATION_INSTRUCTION.search(label.strip()) or
                   (not (node.clickable or node.long_clickable or node.editable) and self.VERIFICATION_HEADING.fullmatch(label.strip()))
                   for node in observation.nodes if not node.password
                   for label in (node.text, node.description))

    def evaluate(self, mode: str, command: Command, observation: Observation | None) -> Verdict:
        if command.kind in {"observe", "screenshot", "wait"}:
            return Verdict("allow", "Read-only observation")
        if self.requires_verification(observation):
            return Verdict("manual", "请在手机上手动完成安全验证，再点击继续。", "verification")
        node = None
        payment = command.kind == "pay"
        if command.payment_consent_id and not payment:
            return Verdict("manual", "旧付款指令已失效，请重新观察后决策", "payment")
        if payment and (mode != "full" or observation is None or not command.payment_consent_id or
                        command.payment_consent_id != observation.payment_consent_id):
            return Verdict("manual", "付款需要完全访问权限和有效的付款授权，请手动处理", "payment")
        if command.kind in {"tap", "pay", "long_press", "type", "login_phone", "login_code"} or command.target:
            if observation is None or command.screen_id != observation.screen_id:
                return Verdict("deny", "Screen changed; observe again before acting")
            node = next((n for n in observation.nodes if n.id == command.target), None)
            if node is None or not node.enabled:
                return Verdict("deny", "Target missing or disabled")
            if node.password:
                return Verdict("manual", "Sensitive input requires manual interaction", "payment" if payment else None)
            if command.kind in {"type", "login_phone", "login_code"} and not node.editable:
                return Verdict("deny", "Target is not editable")
            if command.kind in {"login_phone", "login_code"} and command.package_name != observation.package_name:
                return Verdict("deny", "Local login is scoped to the current application")
            if command.kind in {"tap", "pay"} and not node.clickable:
                return Verdict("deny", "Target is not clickable")
            if command.kind == "long_press" and not node.long_clickable:
                return Verdict("deny", "Target does not support long press")
        if payment:
            return Verdict("allow", "本次付款已取得完全访问与本机付款授权")
        if mode == "ask" and command.kind in {"tap", "pay", "long_press", "type", "login_phone", "login_code", "launch", "open_document", "recents", "notifications", "quick_settings", "split_screen"}:
            return Verdict("approve", "该操作需要你的批准")
        if mode == "assist" and node is not None and command.kind in {"tap", "long_press"}:
            label = f"{node.text} {node.description}".lower().strip()
            if not label or any(word in label for word in self.SENSITIVE_WORDS):
                return Verdict("approve", "请确认这次操作的目标与内容")
        return Verdict("allow", "Within current execution mode")

File: android/sdk/src/main/java/dev/doppel/sdk/Policy.kt
package dev.doppel.sdk

import java.security.MessageDigest

data class Target(val label: String, val password: Boolean, val enabled: Boolean)

object Policy {
    private val payment = Regex("支付|付款|扣款|快付|免密|转账|汇款|\\bpay(?:now)?\\b|\\bpayment\\b|purchase|checkout|transfer", RegexOption.IGNORE_CASE)
    private val manualFinancial = Regex("转账|汇款|\\btransfer\\b|remittance|支付密码|付款密码|支付验证码|付款验证码|(?:开通|开启|设置|授权|关闭|取消).{0,10}(?:免密|自动扣款|自动续费)|(?:免密|自动扣款|自动续费).{0,10}(?:开通|开启|设置|授权|关闭|取消)|payment.{0,12}(?:password|passcode|verification code|otp)|direct debit|standing order|recurring payment", RegexOption.IGNORE_CASE)
    private val verificationHeading = Regex("(?:人机验证|安全验证|安全校验|滑块验证|点选验证|行为验证|图形验证|图形验证码|图片验证码|captcha|recaptcha|hcaptcha|security verification|security check|human verification)[.!。！:：]?", RegexOption.IGNORE_CASE)
    private val verificationInstruction = Regex("(?:请|需要).{0,12}(?:完成|进行).{0,12}(?:安全|人机)验证|请验证[您你]是真人|(?:拖动|滑动).{0,12}滑块.{0,20}(?:验证|拼图)|完成拼图|(?:drag|slide).{0,24}slider.{0,40}(?:verif|puzzle)|(?:select|click) all (?:images|squares) (?:with|containing)|^verify (?:that )?you are (?:a )?human[.!]?$|^i['’]?m not a robot[.!]?$|complete.{0,12}(?:security verification|captcha)", RegexOption.IGNORE_CASE)
    private val registrationTarget = Regex("(?:注册|立即注册|免费注册|新用户|sign\\s*up|register)", RegexOption.IGNORE_CASE)
    private val registrationPage = Regex("(?:注册账号|注册帐户|创建账号|创建帐户|注册新账号|注册新帐户|新用户注册|create\\s+(?:a\\s+)?(?:new\\s+)?account|sign\\s*up|register\\s+(?:an?\\s+)?account)", RegexOption.IGNORE_CASE)
    fun sensitive(label: String) = payment.containsMatchIn(label)
    fun manualFinancial(label: String): Boolean {
        val normalized = Regex("已(?:成功)?(?:开通|开启|授权|关闭|设置|取消)").replace(label, "状态生效")
        val mandate = "auto(?:matic)?[\\s-]*(?:pay(?:ment)?s?|renew(?:al)?|debit|billing)"
        val action = "(?:enable|activate|authorize|set\\s*up|disable|cancel)"
        return manualFinancial.containsMatchIn(normalized) || Regex("\\b$action\\b.{0,20}\\b$mandate\\b|\\b$mandate\\b.{0,20}\\b$action\\b", RegexOption.IGNORE_CASE).containsMatchIn(normalized)
    }
    fun financialCredential(label: String) = Regex("密码|口令|验证码|校验码|动态码|\\b(?:password|passcode|pin|otp)\\b|one.?time|verification.?code", RegexOption.IGNORE_CASE).containsMatchIn(label)
    fun verificationRequired(labels: Iterable<String>) = labels.any { verificationLabel(it, false) }
    /** Strict product rule: the agent may never create a new third-party account. */
    fun registrationTarget(label: String): Boolean = registrationTarget.containsMatchIn(label.take(4000))
    fun registrationContext(labels: Iterable<String>): Boolean = registrationPage.containsMatchIn(labels.joinToString(" ").take(12000))
    fun verificationLabel(label: String, interactive: Boolean): Boolean {
        val text = label.take(4000).trim()
        return verificationInstruction.containsMatchIn(text) || !interactive && verificationHeading.matches(text)
    }
    fun codeInput(label: String) = Regex("验证码|校验码|动态码|one.?time|otp|verification.?code", RegexOption.IGNORE_CASE).containsMatchIn(label)
    fun phoneInput(label: String) = Regex("手机|电话|(?:^|[^a-z])(?:mobile|phone|telephone)(?:[^a-z]|$)", RegexOption.IGNORE_CASE).containsMatchIn(label)
    fun redactLoginLabel(label: String) = Regex("\\p{Nd}").replace(label, "*")
    /** Payment is an explicit planner action; the host checks authority, never page words. */
    fun canPay(mode: String, paymentAuthorized: Boolean, packageName: String) =
        mode == "full" && paymentAuthorized && packageName.isNotBlank() && !packageName.startsWith("window:")
    fun validate(kind: String, expected: String?, actual: String, target: Target?, mode: String = "assist"): String {
        if (mode !in setOf("ask", "assist", "full")) return "blocked"
        if (kind in setOf("tap", "long_press", "type", "login_phone", "login_code", "login_username", "login_password", "scroll")) {
            if (expected.isNullOrBlank() || expected != actual) return "stale"
            if (target == null) return "stale"
            if (target.password && kind != "login_password" && !(kind == "login_code" && codeInput(target.label)) || !target.enabled) return "blocked"
            // Credential field restrictions do not classify ordinary navigation as a payment.
            if (kind in setOf("login_phone", "login_code", "login_username", "login_password") && manualFinancial(target.label)) return "blocked"
        }
        return "ok"
    }
    fun hash(content: String): String = MessageDigest.getInstance("SHA-256").digest(content.toByteArray(Charsets.UTF_8)).joinToString("") { "%02x".format(it) }
}

class CommandLedger(private val read: (String) -> String?, private val write: (String, String) -> Boolean) {
    @Synchronized fun cached(id: String) = read(id)
    @Synchronized fun claim(id: String, uncertainResult: String): Boolean = read(id) == null && write(id, uncertainResult)
    @Synchronized fun finish(id: String, result: String): Boolean = write(id, result)
}
